### ISSUE TITLE
Implement some sort of accessible/text-alternative for callout icons or blocks (a11y issue)

### ISSUE THREAD

**Issue by batpigandme** [quarto-dev/quarto-cli#726]
I'd argue that the icons used in [callout blocks](https://quarto.org/docs/authoring/callouts.html) are semantic icons, _or_ (alternatively) that the blocks themselves are intended as "semantic" classes.

As such, I think there should be visually hidden, but accessible text for either the icons or the callout divs. Because the icons are used as background images, I think that of the [accessibility techniques for semantic icons](https://fontawesome.com/docs/web/dig-deeper/accessibility#semantic-icons) (link is from Font Awesome, but the same ideas apply), we'd probably want to use Bootstrap's [`.visually-hidden`](https://getbootstrap.com/docs/5.1/getting-started/accessibility/#visually-hidden-content) class anyhow—which is why I think it might make sense to add this to the divs, rather than just the icons (since it will then work with `appearance="minimal"` or `icon=false`).

I suppose the downside to this is that you'd only want it to apply to HTML, which might be cumbersome given the callouts are used in other formats. One alternative would be to supply an `aria-label`, though I think screen-reader support for them on things like `<div>`s and `<span>`s can be a bit inconsistent.

**Comment by batpigandme** [comment]
Outside of Bootstrap, `.visually-hidden`, `.sr-only`, etc. are pretty common classes for accessibility. For example, here's a pattern from one of the [Deque University](https://dequeuniversity.com/) accessibility-training courses I've been taking that uses its own `.visually-hidden` to make `<ins>` and `<del>` (which have no semantic meaning to screen readers) accessible:
```html
<head>
    <title>Sample web page>
        <style>
            .visually-hidden {
                position: absolute;
                clip: rect(0 0 0 0);
                border: 0;
                height: 1px;
                margin: -1px;
                overflow: hidden;
                padding: 0 width: 1px;
            }
        </style>
</head>

<body>
    <p>If you can't say something nice, don't say
        <del>
            <span class="visually-hidden">begin deleted text</span>
            nothing
            <span class="visually-hidden">end deleted text</span>
        </del>
        <ins>
            <span class="visually-hidden">begin inserted text</span>
            anything
            <span class="visually-hidden">end inserted text</span>
        </ins>
        at all.
    </p>
</body>
```

For callouts, I manually added a `.visually-hidden` `<span>` inside the notes on a [quarto website page here](https://quartoand.me/css-grid-and-accessibility#css-grid-power-and-responsibility). The html comes out:
```html
<div class="callout-note callout callout-style-simple">
    <div class="callout-body d-flex">
        <div class="callout-icon-container">
            <i class="callout-icon"></i>
        </div>
        <div class="callout-body-container">
            <p><span class="visually-hidden">Note: </span> From here on out, if I use the term “Grid” it will be in
                reference to CSS Grid, despite the fact that other grid systems exist (such as Bootstrap’s native grid).
            </p>
        </div>
    </div>
</div>
```

It would be somewhat more complicated to add the text to the icon, due to the fact that its in the CSS as a `background-image`.

There are other CSS implementations of what are, in effect,`.visually-hidden`, e.g. the [Improved sr-only](https://gist.github.com/ffoodd/000b59f431e3e64e4ce1a24d5bb36034) by @ffoodd ([Code Pen of same](https://codepen.io/ffoodd/pen/gwKeyq).

Deque (makers of axe, which powers most of the popular accessibility tools) has a [Code Library of Accessibility Examples](https://dequeuniversity.com/library/), and their [CSS patterns](https://dequeuniversity.com/assets/js/patterns/deque-patterns.css) uses the implementation below for `.visually-hidden`, which should work for reveal as well:
```css
@charset "UTF-8";

.visuallyhidden {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
```

**Comment by batpigandme** [comment]
For callouts with captions, I couldn't add the `.visually-hidden` span inside of the callout (it messes up the caption rendering), so I added a begin and end point outside of the callout:
```html
<p><span class="visually-hidden">Note: </span></p>
<div class="callout-note callout callout-style-simple callout-captioned">
  <div class="callout-header d-flex align-content-center">
    <div class="callout-icon-container">
      <i class="callout-icon"></i>
    </div>
    <div class="callout-caption-container flex-fill">
      Pro Tip
    </div>
  </div>
  <div class="callout-body-container callout-body">
    <p>Contents of my tip.</p>
  </div>
</div>
<p><span class="visually-hidden">End note.</span></p>
```

Live in [webpage here](https://quartoand.me/front-end.html#css-grid).

**Comment by jjallaire** [comment]
Thank you! We will do something based on what you have provided here!

**Comment by mcanouil** [comment]
FYI, the Quarto Iconify extension allows to set title and label (aria-label): https://github.com/mcanouil/quarto-iconify?tab=readme-ov-file

**Comment by cscheid** [comment]
I like the idea of Quarto providing `.screen-reader-only` as a CSS class for bootstrap (`html`, `dashboard`, etc) and `revealjs` documents. We should document it and then get good at using it ourselves.

(to be clear: using our just-merged `format: html: axe: ...` checker, I can tell that our current callout implementation doesn't produce `axe-core` _violations_. But I agree that it would be good to go the extra mile.)

**Comment by cscheid** [comment]
These are the basic patterns of callout usage:

````
::: callout-note

This is a callout.

:::

::: callout-note

## Title

This is a callout.

:::

::: {#nte-1 .callout-note}

## A title

This is a crossreferenceable callout with a title.

:::

::: {#nte-2 .callout-note}

This is a crossreferenceable callout without a title.

:::
````

They produce the following visual output:

<img width="866" alt="Image" src="https://github.com/user-attachments/assets/c0b5b7d0-f118-488b-9565-e457909c6143" />

Of these cases, the one that needs additional help is 

````
::: callout-note

## Title

This is a callout.

:::
````

Where nothing but the icon indicates it's a Note. So that's the one we'll use `.screen-reader-only` for.

**Comment by cscheid** [comment]
This is how the HTML will look like now:

````
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Title
</div>
</div>
<div class="callout-body-container callout-body">
<p>This is a callout.</p>
</div>
</div>
````
